### PR TITLE
REP-4653 UAE Lost Password

### PR DESCRIPTION
**Pull request by Mario-Lopez**
Story is here:
https://repose.atlassian.net/browse/REP-4653

Updated Rackspace Auth User filter to read the request body of a Forgot Password request to grab the username, and it will only attempt to do so if the URL matches the current one used by Identity (trailing slash optional).  Otherwise, it will act like normal and try to read the request body like it was an auth request.  All of this only applies if it's a POST request, of course.

Also updated HERP filter to check the response from the origin service for an X-User-Name header if there wasn't one on the request.

### CODE REVIEW

File: repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/filters/herp/HerpUserAccessEventFilterTest.groovy [hunk 1]
```diff
@@ -446,53 +443,109 @@ class HerpUserAccessEventFilterTest extends ReposeValveTest {
         "200"        | "test12" | "tenantId=12345" | "PUT"  | "OK"
     }
 
+    @Unroll
+    def "will populate and log the username '#expectedUsername' when the request header username is '#requestUsername' and the response header is '#responseUsername'"() {
+        given:
+        def requestHeaders = requestUsername ? [(OpenStackServiceHeader.USER_NAME.toString()): requestUsername] : [:]
+        def responseHeaders = responseUsername ? [(OpenStackServiceHeader.USER_NAME.toString()): responseUsername] : [:]
```
Comment (Mario-Lopez) [inline repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/filters/herp/HerpUserAccessEventFilterTest.groovy]: This needs to be an explicit null check since I test empty strings which are "falsy" in Groovy.

File: repose-aggregator/components/filters/rackspace-auth-user-filter/src/main/scala/org/openrepose/filters/rackspaceauthuser/RackspaceAuthUserFilter.scala [hunk 2]
```diff
@@ -75,65 +77,50 @@ class RackspaceAuthUserFilter @Inject()(configurationService: ConfigurationServi
...
   }
 
-  def getUsername(domain: String, user: String): String = {
-    if (domain.equals("Rackspace")) {
-      "Racker:" + user
-    } else {
-      user
-    }
-  }
+  def getUsername(domain: String, user: String): String = if (domain.equals("Rackspace")) s"Racker:$user" else user
```
Comment (adrianjgeorge) [inline repose-aggregator/components/filters/rackspace-auth-user-filter/src/main/scala/org/openrepose/filters/rackspaceauthuser/RackspaceAuthUserFilter.scala]: Not sure i like the collapse here. Probably because i don't like the lack of a ternary operator in scala. This is one of those things that is a bug waiting to happen.

File: repose-aggregator/components/filters/rackspace-auth-user-filter/src/main/scala/org/openrepose/filters/rackspaceauthuser/RackspaceAuthUserFilter.scala [hunk 3]
```diff
@@ -187,12 +173,33 @@ class RackspaceAuthUserFilter @Inject()(configurationService: ConfigurationServi
...
+      case failure: JsError =>
+        logger.debug("2.0 JSON parsing failure: {}", JsError.toFlatJson(failure))
+        (None, None)
+      case _ => (None, None)
+    }
   }
 
+  def parseUserGroupFromInputStream(uri: String, inputStream: InputStream, contentType: String, sessionIds: List[String]): Option[RackspaceAuthUserGroup] =
+    uri match {
+      case "/v2.0/users/RAX-AUTH/forgot-pwd" => Option(configuration.getV20).flatMap(parseUsername(_, inputStream, contentType, usernameForgotPassword2_0Json, usernameForgotPassword2_0Xml))
```
Comment (adrianjgeorge) [inline repose-aggregator/components/filters/rackspace-auth-user-filter/src/main/scala/org/openrepose/filters/rackspaceauthuser/RackspaceAuthUserFilter.scala]: will this ever have a trailing slash?

File: repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/filters/herp/HerpUserAccessEventFilterTest.groovy (hunk 1 above)
Comment (Mario-Lopez) [inline repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/filters/herp/HerpUserAccessEventFilterTest.groovy]: I fixed.

File: repose-aggregator/components/filters/rackspace-auth-user-filter/src/main/scala/org/openrepose/filters/rackspaceauthuser/RackspaceAuthUserFilter.scala (hunk 2 above)
Comment (Mario-Lopez) [inline repose-aggregator/components/filters/rackspace-auth-user-filter/src/main/scala/org/openrepose/filters/rackspaceauthuser/RackspaceAuthUserFilter.scala]: I will uncollapse.

File: repose-aggregator/components/filters/rackspace-auth-user-filter/src/main/scala/org/openrepose/filters/rackspaceauthuser/RackspaceAuthUserFilter.scala (hunk 3 above)
Comment (Mario-Lopez) [inline repose-aggregator/components/filters/rackspace-auth-user-filter/src/main/scala/org/openrepose/filters/rackspaceauthuser/RackspaceAuthUserFilter.scala]: Identity staging works if I include a trailing slash in the URL, so I'll add a functional test for it and update this line to handle it.

File: repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/filters/rackspaceauthuser/RackspaceAuthUserTest.groovy [hunk 4]
```diff
@@ -194,4 +194,72 @@ class RackspaceAuthUserTest extends ReposeValveTest {
         and: "The result will be passed through"
         messageChain.receivedResponse.code == "200"
     }
+
+    @Unroll("Will parse Forgot Password request for username when sent to appropriate URL (#testName)")
+    def "Will parse a Forgot Password request if the URL matches /v2.0/users/RAX-AUTH/forgot-pwd"() {
+        when: "Request body contains forgot password credentials"
+        def messageChain = deproxy.makeRequest([url: reposeEndpoint + "/v2.0/users/RAX-AUTH/forgot-pwd", requestBody: requestBody, headers: contentType, method: "POST"])
+        def sentRequest = ((MessageChain) messageChain).getHandlings()[0]
```
Comment (Mario-Lopez) [inline repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/filters/rackspaceauthuser/RackspaceAuthUserTest.groovy]: I assume Bill copy/pasta'd this from the previous lines which did this all over.  Why this was done before is... well... the old Groovy code isn't that great.  I will clean some of this mess up.

File: repose-aggregator/src/docs/asciidoc/filters/herp.adoc [hunk 5]
```diff
@@ -16,7 +16,17 @@ Events are recorded even if the API request is rejected by Repose before reachin
 
 == Prerequisites & Postconditions
 === Required Headers
-This filter has no required headers.
+[cols="15,12,73", options="header"]
```
Comment (adrianjgeorge) [inline repose-aggregator/src/docs/asciidoc/filters/herp.adoc]: Why are you specifying the column sizes?

File: repose-aggregator/src/docs/asciidoc/filters/rackspace-auth-user.adoc [hunk 6]
```diff
@@ -38,11 +38,11 @@ It's meant to be used for login requests to capture username for both <<../recip
...
 |===
 
-=== Required preceding filters
+=== Required Preceding Filters
 This filter has no dependencies on other filters.
 
-=== Request headers created
-[cols="3", options="header,autowidth"]
+=== Request Headers Created
+[cols="15,12,73", options="header"]
```
Comment (adrianjgeorge) [inline repose-aggregator/src/docs/asciidoc/filters/rackspace-auth-user.adoc]: Sized here to for some reason?

File: repose-aggregator/src/docs/asciidoc/filters/herp.adoc (hunk 5 above)
Comment (Mario-Lopez) [inline repose-aggregator/src/docs/asciidoc/filters/herp.adoc]: It looked pretty ugly with `autowidth`, mostly because of the header names.  The first two columns were squished and required wrapping.  I thought it looked better this way.  I can always change back if not better?
